File: batches_c/batch_13_different_implementation_same_algorithm/ref.c
#include <stdio.h>

int binary_search(int *arr, int len, int target) {
    int left = 0, right = len - 1;
    while (left <= right) {
        int center = left + (right - left) / 2;
        if (arr[center] == target) return center;
        if (arr[center] < target) left = center + 1;
        else right = center - 1;
    }
    return -1;
}

int main() {
    int a[] = {1, 3, 5, 7, 9};
    printf("%d\n%d\n", binary_search(a, 5, 5), binary_search(a, 5, 4));
    return 0;
}

File: batches/batches/batch_13_different_implementation_same_algorithm/ref.py
"""
Reference - Original code for Batch 13: Different Implementation Same Algorithm (NOT cheating)
Binary search - verbose style.
"""


def binary_search(arr, target):
    lo = 0
    hi = len(arr) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            return mid
        if arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1


if __name__ == "__main__":
    a = [1, 3, 5, 7, 9]
    print(binary_search(a, 5))
    print(binary_search(a, 4))
